hibernate-core/src/main/java/org/hibernate/query/internal/QueryInterpretationCacheStandardImpl.java

```diff
@@ -101,14 +101,23 @@ public void cacheNonSelectQueryPlan(Key key, NonSelectQueryPlan plan) {
 	@Override
 	public HqlInterpretation resolveHqlInterpretation(
 			String queryString,
+			Class<?> expectedResultType,
 			Function<String, SqmStatement<?>> creator) {
 		log.tracef( "QueryPlan#resolveHqlInterpretation( `%s` )", queryString );
 		final StatisticsImplementor statistics = statisticsSupplier.get();
 		final boolean stats = statistics.isStatisticsEnabled();
 		final long startTime = ( stats ) ? System.nanoTime() : 0L;
 
+		final String cacheKey;
+		if ( expectedResultType != null && expectedResultType.isArray() ) {
+			cacheKey = queryString + "_array";
```

Comment:
As discussed, prepending `"array_"` would be better to avoid possible collisions.